Dave

abelnation opened the pull request:
@dpekar 
making a pull request so i can make comments

dpekar committed:
Player constructor and basic tests

dpekar committed:
Tile and GameSate models + test files

dpekar committed:
Added investment markets to player model

dpekar committed:
GameState creates set of players on instantiation

dpekar committed:
GameState initialization now sets up tile piles

dpekar committed:
Setup of starting tiles for players

dpekar committed:
Adding placement to mode manager

dpekar committed:
Fixed bug in iterator when adding tiles to GameState

dpekar committed:
Setup market in gamestate initialization

dpekar committed:
Moved GameSetup functionality into mixin

dpekar committed:
Blocked out functionality needde in GameState

dpekar committed:
All basic game objects and specs in!

abelnation commented on src/shared/data/Tile-list.es6:
```diff
@@ -0,0 +1,183 @@
+
+module.exports = {
+
+    'basic': [
```
We might want to make each of these an object instead of an array so that we can grab specific tiles by the id or something similar.  There may be cases where we need to check for specific tile types?  Not sure, just a thought

dpekar replied:
Hmm, maybe but I'm not quite sure what you have in mind.  I'm going to keep it this way for now but let's discuss live and happy to change :)

abelnation commented on src/shared/data/Tile-list.es6:
```diff
@@ -0,0 +1,183 @@
+        {
+          name: 'Suburbs',
+          cost: 3,
+          category: 'Residential',
```
When using string constants like this, we should define them in a class or module.  e.g.:

```
const TileCategories = require('...')
...
  {
    name: 'Suburbs',
    cost: 3,
    category: TileCategories.RESIDENTIAL,
    icon: TileIcons.SCHOOL,
    stage: GameStages.BASIC,
    ...
  }
```

dpekar replied:
Added Tile.CATEGORIES constant to Tile class.

dpekar replied:
Actually, now there is a TileConfig module. So TileConfig.CATEGORIES

abelnation commented on src/shared/data/Tile-list.es6:
```diff
@@ -0,0 +1,183 @@
+        }
+    ],
+
+    'a': [
```
`'a', 'b', 'c'` can also be a constant

abelnation commented on src/shared/models/game/GameSetup.es6:
```diff
@@ -0,0 +1,83 @@
+const _ = require('underscore')
+
+
+const GameSetup = {
```
So will this represent just the starting state? or is the gamesetup used to represent the board throughout the game?

dpekar replied:
Yeah so I mentioned this in chat, but GameState is the representation of state that we'll be keeping and using throughout the game.

GameSetup is just a mixin for setting up the initial game state before any turns have been made to keep things neat / separate out the diff concerns.

abelnation commented on src/shared/models/game/GameState.es6:
```diff
@@ -0,0 +1,58 @@
+
+
+    placeTile(tile, player) {
+        // remove a tile from the market or basic market
```
as you think about these steps.  we'll probably want to make each step something we can call separately.  i can see, in the client, us wanting to play out each step one at a time

dpekar replied:
yes absolutely,  I have a design in mind for this.  I put it in the eng design doc last night.  I think every turn should get back a result object which is basically an array of effects that can be played out in order.  Each "effect" holds an stat (income), the change (+2), and the placement (farm on (2,2)) so it can all be animated on the front end step by step.

abelnation commented on src/shared/models/game/Market.es6:
```diff
@@ -0,0 +1,45 @@
+    getTiles() { return this.get('tiles') }
+
+
+    fillUpSlots(tilePiles) {
```
i assume this class will also have a `replenishMarket`, or some similar method for adding a single tile? also `trashTileInMarket`

dpekar replied:
Yep! Hoping to get to some of this today, but maybe not till Monday

abelnation commented on src/shared/models/game/Market.es6:
```diff
@@ -0,0 +1,45 @@
+        for (let stage in Tile.STAGES) {
+            let pile = tilePiles[Tile.STAGES[stage]]
+            if (pile.length > 0) {
+                return pile.shift() 
```
and then `break` out of the for loop?

dpekar replied:
Does this need an explicit break?  I thought the return suffices?

abelnation commented on src/shared/models/game/Market.es6:
```diff
@@ -0,0 +1,45 @@
+}
+
+Market.SLOT_COSTS = [10, 8, 6, 4, 2, 0, 0]
+Market.SLOTS = Market.SLOT_COSTS.length
```
can we call this `NUM_SLOTS`?

dpekar replied:
good point!  done.

abelnation commented on src/shared/models/game/Placement.es6:
```diff
@@ -0,0 +1,29 @@
+const BaseModel = require('../BaseModel')
+
+class Placement extends BaseModel {
+    constructor(tile, coordinates, turn) {
```
what does `turn` mean in this context?

dpekar replied:
It's the number of turns into the game this tile was placed on.  Might be useful for things like Restaurants, SkyScraper. 

Changed to turnNum. To be more clear.

abelnation replied:
Gotcha

abelnation commented on src/shared/models/game/Tile.es6:
```diff
@@ -0,0 +1,58 @@
+class Tile extends BaseModel {
+    constructor(tileInfo) {
+        super()
+        this.set('name', tileInfo.name)
```
maybe it'd be worth creating a method in BaseModel, `setFromObject`, where:

```
var x = {
  a: 1,
  b: 2
}

model.setFromObject(x)
--> would call -->
model.set('a', 1)
model.set('b', 2)
```

dpekar replied:
Good call.

abelnation commented on src/shared/data/Tile-list.es6:
```diff
@@ -0,0 +1,183 @@
+          category: 'Industrial',
+          icon: null,
+          stage: 'c',
+          immediateEffect: (player) => { player.income = player.income + 2 },
```
i think these immediate and conditional effects will probably need to be passed the entire gameState so that they can examine other player boards as well.  perhaps they just need to be passed all players

dpekar replied:
Yeah I think they just need players, not gamestate.  B/c they don't need to know what's in the market etc.

Added.

dpekar committed:
Need all players to execute conditional effects of tiles

dpekar committed:
Factored out config constats into their own modules

dpekar committed:
Removed unused variables from all files

dpekar committed:
Added setFromObject to BaseModel for use in constructors

dpekar committed:
Added board model & support for lakes

dpekar committed:
Added testing for charging a player for buying tile

dpekar committed:
Added canAfford function to player model

dpekar committed:
Added support for immedate effects

dpekar committed:
Basic implementation

Still need testing for every, other, etc various conditionals.

dpekar committed:
All basic effects working

dpekar committed:
Moved effect execution out of player model into mixin

dpekar committed:
GameSetup isn't a model so removed from ModelManager

dpekar committed:
Update a player's population and money at end of turn

dpekar committed:
Basic turn complete!

dpekar committed:
Fixed bug with Tile copies not being true TIiles

dpekar committed:
All turns types implemented and tested lightly :D

dpekar commented on src/shared/models/game/Effect.es6:
```diff
@@ -0,0 +1,106 @@
+        }
+    }
+
+    executeOn(player, placement, gameState) {
```
I need to create subclasses for effects to get rid of this nasty switch statement.

dpekar committed:
Small fixes to investments

dpekar committed:
Fixed resolution of adjacent tlies

dpekar committed:
Turns now produce array of EffectResults

An EffectResult summarizes one change to the gameState via a combination of placement (what placement caused it to change), stat (what stat is affeceted) and value (how much did the stat change).

dpekar committed:
Cleanup for commit
